File: clients/javascript/leafmind-client.js
/**
 * LeafMind JavaScript/Node.js Client SDK
 * 
 * A neuromorphic memory database client for JavaScript and TypeScript applications.
 */

import fetch from 'node-fetch';

/**
 * Configuration for memory recall operations
 */
export class RecallQuery {
    constructor({
        maxResults = 10,
        minRelevance = 0.1,
        useRecencyBoost = true,
        explorationBreadth = 5,
        maxPathLength = 3
    } = {}) {
        this.maxResults = maxResults;
        this.minRelevance = minRelevance;
        this.useRecencyBoost = useRecencyBoost;
        this.explorationBreadth = explorationBreadth;
        this.maxPathLength = maxPathLength;
    }
}

/**
 * Represents a concept in the LeafMind memory system
 */
export class Concept {
    constructor(data) {
        this.id = data.id;
        this.content = data.content;
        this.metadata = data.metadata || {};
        this.createdAt = data.created_at;
        this.lastAccessed = data.last_accessed;
        this.accessCount = data.access_count;
    }
}

/**
 * Represents a recall result with relevance score
 */
export class RecallResult {
    constructor(data) {
        this.concept = new Concept(data.concept);
        this.relevanceScore = data.relevance_score;
        this.pathLength = data.path_length;
        this.connectionStrength = data.connection_strength;
    }
}

/**
 * Custom error class for LeafMind API errors
 */
export class LeafMindError extends Error {
    constructor(message, status = null) {
        super(message);
        this.name = 'LeafMindError';
        this.status = status;
    }
}

/**
 * LeafMind Client for JavaScript/Node.js
 * 
 * This client provides a JavaScript interface to interact with a LeafMind API server,
 * allowing you to use LeafMind as a database from your JavaScript/Node.js applications.
 * 
 * @example
 * ```javascript
 * const client = new LeafMindClient('http://localhost:8080');
 * 
 * // Learn new concepts
 * const conceptId = await client.learn('JavaScript is a programming language');
 * 
 * // Create associations
 * const otherId = await client.learn('Programming requires logical thinking');
 * await client.associate(conceptId, otherId);
 * 
 * // Recall related memories
 * const results = await client.recallFromConcept(conceptId);
 * results.forEach(result => {
 *     console.log(`Recalled: ${result.concept.content} (score: ${result.relevanceScore})`);
 * });
 * ```
 */
export class LeafMindClient {
    /**
     * Initialize the LeafMind client
     * 
     * @param {string} baseUrl - The base URL of the LeafMind API server
     * @param {string} [apiKey] - Optional API key for authentication
     */
    constructor(baseUrl = 'http://localhost:8080', apiKey = null) {
        this.baseUrl = baseUrl.replace(/\/+$/, '');
        this.apiKey = apiKey;
        this.headers = {
            'Content-Type': 'application/json',
        };
        
        if (this.apiKey) {
            this.headers['Authorization'] = `Bearer ${this.apiKey}`;
        }
    }

    /**
     * Make an HTTP request to the API
     * @private
     */
    async _request(method, endpoint, data = null) {
        const url = `${this.baseUrl}${endpoint}`;
        const options = {
            method,
            headers: this.headers,
        };

        if (data !== null) {
            options.body = JSON.stringify(data);
        }

        try {
            const response = await fetch(url, options);
            
            if (!response.ok) {
                const errorText = await response.text();
                let errorMessage = `HTTP ${response.status}: ${response.statusText}`;
                
                try {
                    const errorJson = JSON.parse(errorText);
                    errorMessage = errorJson.error || errorMessage;
                } catch {
                    // Use default error message if JSON parsing fails
                }
                
                throw new LeafMindError(errorMessage, response.status);
            }

            // Handle empty responses (like 204 No Content)
            if (response.status === 204) {
                return null;
            }

            return await response.json();
        } catch (error) {
            if (error instanceof LeafMindError) {
                throw error;
            }
            throw new LeafMindError(`Connection error: ${error.message}`);
        }
    }

    // Health and Status
    
    /**
     * Check the health status of the LeafMind server
     * @returns {Promise<Object>} Health status information
     */
    async healthCheck() {
        return await this._request('GET', '/health');
    }

    /**
     * Get memory system statistics
     * @returns {Promise<Object>} Memory statistics
     */
    async getStats() {
        return await this._request('GET', '/stats');
    }

    // Core Memory Operations

    /**
     * Learn a new concept
     * 
     * @param {string} content - The content of the concept to learn
     * @param {Object} [metadata] - Optional metadata object
     * @returns {Promise<string>} The UUID of the newly created concept
     */
    async learn(content, metadata = null) {
        const requestData = {
            content: content,
            metadata: metadata
        };
        const response = await this._request('POST', '/concepts', requestData);
        return response.concept_id;
    }

    /**
     * Retrieve a specific concept by ID
     * 
     * @param {string} conceptId - The UUID of the concept to retrieve
     * @returns {Promise<Concept|null>} The concept if found, null otherwise
     */
    async getConcept(conceptId) {
        try {
            const response = await this._request('GET', `/concepts/${conceptId}`);
            if (response?.concept) {
                return new Concept(response.concept);
            }
            return null;
        } catch (error) {
            if (error.status === 404) {
                return null;
            }
            throw error;
        }
    }

    /**
     * List all concepts with pagination
     * 
     * @param {number} [page=1] - Page number (1-indexed)
     * @param {number} [pageSize=50] - Number of concepts per page
     * @returns {Promise<Object>} Object with concepts list and pagination info
     */
    async listConcepts(page = 1, pageSize = 50) {
        return await this._request('GET', `/concepts?page=${page}&page_size=${pageSize}`);
    }

    /**
     * Delete a concept
     * 
     * @param {string} conceptId - The UUID of the concept to delete
     * @returns {Promise<boolean>} True if deleted successfully
     */
    async deleteConcept(conceptId) {
        try {
            await this._request('DELETE', `/concepts/${conceptId}`);
            return true;
        } catch (error) {
            if (error.status === 404) {
                return false;
            }
            throw error;
        }
    }

    /**
     * Access a concept (update access time and count)
     * 
     * @param {string} conceptId - The UUID of the concept to access
     * @returns {Promise<boolean>} True if accessed successfully
     */
    async accessConcept(conceptId) {
        try {
            await this._request('POST', `/concepts/${conceptId}/access`);
            return true;
        } catch (error) {
            if (error.status === 404) {
                return false;
            }
            throw error;
        }
    }

    // Association Operations

    /**
     * Create an association between two concepts
     * 
     * @param {string} fromConceptId - Source concept UUID
     * @param {string} toConceptId - Target concept UUID
     * @param {boolean} [bidirectional=false] - Whether to create a bidirectional association
     * @returns {Promise<boolean>} True if association created successfully
     */
    async associate(fromConceptId, toConceptId, bidirectional = false) {
        const requestData = {
            from_concept_id: fromConceptId,
            to_concept_id: toConceptId,
            bidirectional: bidirectional
        };
        const response = await this._request('POST', '/associations', requestData);
        return response?.edge_created || false;
    }

    /**
     * Remove an association between two concepts
     * 
     * @param {string} fromConceptId - Source concept UUID
     * @param {string} toConceptId - Target concept UUID
     * @returns {Promise<boolean>} True if association removed successfully
     */
    async removeAssociation(fromConceptId, toConceptId) {
        try {
            await this._request('DELETE', `/associations/${fromConceptId}/${toConceptId}`);
            return true;
        } catch (error) {
            if (error.status === 404) {
                return false;
            }
            throw error;
        }
    }

    // Recall Operations

    /**
     * Recall memories starting from a specific concept
     * 
     * @param {string} conceptId - The UUID of the source concept
     * @param {RecallQuery} [query] - Optional recall query configuration
     * @returns {Promise<RecallResult[]>} Array of recall results
     */
    async recallFromConcept(conceptId, query = null) {
        if (!query) {
            query = new RecallQuery();
        }

        const requestData = {
            source_concept_id: conceptId,
            max_results: query.maxResults,
            min_relevance: query.minRelevance,
            use_recency_boost: query.useRecencyBoost,
            exploration_breadth: query.explorationBreadth,
            max_path_length: query.maxPathLength
        };

        const response = await this._request('POST', '/recall', requestData);
        return response.results.map(result => new RecallResult(result));
    }

    /**
     * Recall memories by content similarity
     * 
     * @param {string} contentQuery - Text to search for
     * @param {RecallQuery} [query] - Optional recall query configuration
     * @returns {Promise<RecallResult[]>} Array of recall results
     */
    async recallByContent(contentQuery, query = null) {
        if (!query) {
            query = new RecallQuery();
        }

        const requestData = {
            content_query: contentQuery,
            max_results: query.maxResults,
            min_relevance: query.minRelevance,
            use_recency_boost: query.useRecencyBoost,
            exploration_breadth: query.explorationBreadth,
            max_path_length: query.maxPathLength
        };

        const response = await this._request('POST', '/recall/content', requestData);
        return response.results.map(result => new RecallResult(result));
    }

    /**
     * Perform spreading activation recall from a concept
     * 
     * @param {string} conceptId - The UUID of the source concept
     * @returns {Promise<RecallResult[]>} Array of recall results
     */
    async spreadingActivationRecall(conceptId) {
        const requestData = { source_concept_id: conceptId };
        const response = await this._request('POST', '/recall/spreading', requestData);
        return response.results.map(result => new RecallResult(result));
    }

    // Memory Management

    /**
     * Consolidate memory (move short-term to long-term)
     * @returns {Promise<Object>} Consolidation result
     */
    async consolidateMemory() {
        return await this._request('POST', '/memory/consolidate');
    }

    /**
     * Forget weak memories (cleanup)
     * @returns {Promise<Object>} Forgetting result
     */
    async forgetMemories() {
        return await this._request('POST', '/memory/forget');
    }

    /**
     * Optimize memory (consolidate + forget)
     * @returns {Promise<Object>} Optimization result
     */
    async optimizeMemory() {
        return await this._request('POST', '/memory/optimize');
    }

    // Persistence Operations

    /**
     * Save current memory state to disk
     * @returns {Promise<Object>} Save result
     */
    async saveToDisk() {
        const requestData = { action: 'Save' };
        return await this._request('POST', '/persistence', requestData);
    }

    /**
     * Load memory state from disk
     * @returns {Promise<Object>} Load result
     */
    async loadFromDisk() {
        const requestData = { action: 'Load' };
        return await this._request('POST', '/persistence', requestData);
    }

    /**
     * Create a backup of the database
     * @param {string} backupPath - Path for the backup
     * @returns {Promise<Object>} Backup result
     */
    async backupDatabase(backupPath) {
        const requestData = { action: { Backup: { path: backupPath } } };
        return await this._request('POST', '/persistence', requestData);
    }

    /**
     * Restore database from a backup
     * @param {string} backupPath - Path to the backup
     * @returns {Promise<Object>} Restore result
     */
    async restoreDatabase(backupPath) {
        const requestData = { action: { Restore: { path: backupPath } } };
        return await this._request('POST', '/persistence', requestData);
    }

    /**
     * Optimize database storage
     * @returns {Promise<Object>} Optimization result
     */
    async optimizeStorage() {
        const requestData = { action: 'Optimize' };
        return await this._request('POST', '/persistence', requestData);
    }

    /**
     * Get persistence statistics
     * @returns {Promise<Object|null>} Persistence statistics
     */
    async getPersistenceStats() {
        return await this._request('GET', '/persistence/stats');
    }

    // Batch Operations

    /**
     * Learn multiple concepts in a batch
     * 
     * @param {Array<Object>} concepts - Array of concept objects with 'content' and optional 'metadata'
     * @returns {Promise<Array<Object>>} Array of responses for each concept
     */
    async batchLearn(concepts) {
        return await this._request('POST', '/batch/learn', concepts);
    }

    /**
     * Create multiple associations in a batch
     * 
     * @param {Array<Object>} associations - Array of association objects
     * @returns {Promise<Array<Object>>} Array of responses for each association
     */
    async batchAssociate(associations) {
        return await this._request('POST', '/batch/associate', associations);
    }
}

// Convenience functions

/**
 * Quickly learn multiple concepts and optionally create associations between them
 * 
 * @param {LeafMindClient} client - LeafMind client instance
 * @param {string[]} concepts - Array of concept contents to learn
 * @param {boolean} [createAssociations=true] - Whether to create bidirectional associations between all concepts
 * @returns {Promise<string[]>} Array of concept IDs
 */
export async function quickLearnAndAssociate(client, concepts, createAssociations = true) {
    // Learn all concepts
    const conceptIds = [];
    for (const content of concepts) {
        const conceptId = await client.learn(content);
        conceptIds.push(conceptId);
    }

    // Create associations if requested
    if (createAssociations && conceptIds.length > 1) {
        for (let i = 0; i < conceptIds.length; i++) {
            for (let j = 0; j < conceptIds.length; j++) {
                if (i !== j) {
                    await client.associate(conceptIds[i], conceptIds[j], false);
                }
            }
        }
    }

    return conceptIds;
}

/**
 * Build a knowledge graph from concept pairs
 * 
 * @param {LeafMindClient} client - LeafMind client instance
 * @param {Array<[string, string]>} knowledgePairs - Array of [concept1, concept2] pairs to associate
 * @returns {Promise<Object>} Object mapping concept content to concept IDs
 */
export async function buildKnowledgeGraph(client, knowledgePairs) {
    const conceptMap = new Map();

    // Learn all unique concepts
    const allConcepts = new Set();
    for (const [concept1, concept2] of knowledgePairs) {
        allConcepts.add(concept1);
        allConcepts.add(concept2);
    }

    for (const concept of allConcepts) {
        const conceptId = await client.learn(concept);
        conceptMap.set(concept, conceptId);
    }

    // Create associations
    for (const [concept1, concept2] of knowledgePairs) {
        await client.associate(
            conceptMap.get(concept1),
            conceptMap.get(concept2),
            true // bidirectional
        );
    }

    return Object.fromEntries(conceptMap);
}

File: clients/python/leafmind_client.py
"""
LeafMind Python Client SDK

A neuromorphic memory database client for Python applications.
"""

import aiohttp
import asyncio
import json
from typing import Dict, List, Optional, Any, Union
from dataclasses import dataclass
from datetime import datetime
import uuid


@dataclass
class Concept:
    """Represents a concept in the LeafMind memory system."""
    id: str
    content: str
    metadata: Dict[str, str]
    created_at: str
    last_accessed: str
    access_count: int


@dataclass
class RecallResult:
    """Represents a recall result with relevance score."""
    concept: Concept
    relevance_score: float
    path_length: int
    connection_strength: float


@dataclass
class RecallQuery:
    """Configuration for memory recall operations."""
    max_results: int = 10
    min_relevance: float = 0.1
    use_recency_boost: bool = True
    exploration_breadth: int = 5
    max_path_length: int = 3


@dataclass
class MemoryStats:
    """Memory system statistics."""
    total_concepts: int
    short_term_connections: int
    long_term_connections: int
    working_memory_size: int
    last_consolidation: str


class LeafMindClient:
    """
    Async client for LeafMind neuromorphic memory database.
    
    This client provides a Python interface to interact with a LeafMind API server,
    allowing you to use LeafMind as a database from your Python applications.
    
    Example:
        ```python
        async with LeafMindClient("http://localhost:8080") as client:
            # Learn new concepts
            concept_id = await client.learn("Python is a programming language")
            
            # Create associations
            other_id = await client.learn("Programming requires logical thinking")
            await client.associate(concept_id, other_id)
            
            # Recall related memories
            results = await client.recall_from_concept(concept_id)
            for result in results:
                print(f"Recalled: {result.concept.content} (score: {result.relevance_score})")
        ```
    """
    
    def __init__(self, base_url: str = "http://localhost:8080", api_key: Optional[str] = None):
        """
        Initialize the LeafMind client.
        
        Args:
            base_url: The base URL of the LeafMind API server
            api_key: Optional API key for authentication
        """
        self.base_url = base_url.rstrip('/')
        self.api_key = api_key
        self.session: Optional[aiohttp.ClientSession] = None
        
    async def __aenter__(self):
        """Async context manager entry."""
        headers = {}
        if self.api_key:
            headers['Authorization'] = f'Bearer {self.api_key}'
            
        connector = aiohttp.TCPConnector(limit=100)
        timeout = aiohttp.ClientTimeout(total=30)
        
        self.session = aiohttp.ClientSession(
            headers=headers,
            connector=connector,
            timeout=timeout
        )
        return self
        
    async def __aexit__(self, exc_type, exc_val, exc_tb):
        """Async context manager exit."""
        if self.session:
            await self.session.close()
            
    async def _request(self, method: str, endpoint: str, data: Any = None) -> Dict[str, Any]:
        """Make an HTTP request to the API."""
        if not self.session:
            raise RuntimeError("Client not initialized. Use 'async with' context manager.")
            
        url = f"{self.base_url}{endpoint}"
        
        try:
            if data is not None:
                async with self.session.request(method, url, json=data) as response:
                    response.raise_for_status()
                    return await response.json()
            else:
                async with self.session.request(method, url) as response:
                    response.raise_for_status()
                    return await response.json()
        except aiohttp.ClientResponseError as e:
            if e.status == 404:
                raise LeafMindError(f"Resource not found: {e.message}")
            elif e.status == 400:
                raise LeafMindError(f"Bad request: {e.message}")
            elif e.status == 500:
                raise LeafMindError(f"Server error: {e.message}")
            else:
                raise LeafMindError(f"HTTP {e.status}: {e.message}")
        except aiohttp.ClientError as e:
            raise LeafMindError(f"Connection error: {str(e)}")
    
    # Health and Status
    async def health_check(self) -> Dict[str, Any]:
        """Check the health status of the LeafMind server."""
        return await self._request("GET", "/health")
    
    async def get_stats(self) -> MemoryStats:
        """Get memory system statistics."""
        data = await self._request("GET", "/stats")
        return MemoryStats(**data)
    
    # Core Memory Operations
    async def learn(self, content: str, metadata: Optional[Dict[str, str]] = None) -> str:
        """
        Learn a new concept.
        
        Args:
            content: The content of the concept to learn
            metadata: Optional metadata dictionary
            
        Returns:
            The UUID of the newly created concept
        """
        request_data = {
            "content": content,
            "metadata": metadata
        }
        response = await self._request("POST", "/concepts", request_data)
        return response["concept_id"]
    
    async def get_concept(self, concept_id: str) -> Optional[Concept]:
        """
        Retrieve a specific concept by ID.
        
        Args:
            concept_id: The UUID of the concept to retrieve
            
        Returns:
            The concept if found, None otherwise
        """
        try:
            response = await self._request("GET", f"/concepts/{concept_id}")
            if response.get("concept"):
                concept_data = response["concept"]
                return Concept(**concept_data)
            return None
        except LeafMindError as e:
            if "not found" in str(e).lower():
                return None
            raise
    
    async def list_concepts(self, page: int = 1, page_size: int = 50) -> Dict[str, Any]:
        """
        List all concepts with pagination.
        
        Args:
            page: Page number (1-indexed)
            page_size: Number of concepts per page
            
        Returns:
            Dictionary with concepts list, pagination info
        """
        params = {"page": page, "page_size": page_size}
        return await self._request("GET", f"/concepts?page={page}&page_size={page_size}")
    
    async def delete_concept(self, concept_id: str) -> bool:
        """
        Delete a concept.
        
        Args:
            concept_id: The UUID of the concept to delete
            
        Returns:
            True if deleted successfully
        """
        try:
            await self._request("DELETE", f"/concepts/{concept_id}")
            return True
        except LeafMindError as e:
            if "not found" in str(e).lower():
                return False
            raise
    
    async def access_concept(self, concept_id: str) -> bool:
        """
        Access a concept (update access time and count).
        
        Args:
            concept_id: The UUID of the concept to access
            
        Returns:
            True if accessed successfully
        """
        try:
            await self._request("POST", f"/concepts/{concept_id}/access")
            return True
        except LeafMindError as e:
            if "not found" in str(e).lower():
                return False
            raise
    
    # Association Operations
    async def associate(self, from_concept_id: str, to_concept_id: str, bidirectional: bool = False) -> bool:
        """
        Create an association between two concepts.
        
        Args:
            from_concept_id: Source concept UUID
            to_concept_id: Target concept UUID
            bidirectional: Whether to create a bidirectional association
            
        Returns:
            True if association created successfully
        """
        request_data = {
            "from_concept_id": from_concept_id,
            "to_concept_id": to_concept_id,
            "bidirectional": bidirectional
        }
        response = await self._request("POST", "/associations", request_data)
        return response.get("edge_created", False)
    
    async def remove_association(self, from_concept_id: str, to_concept_id: str) -> bool:
        """
        Remove an association between two concepts.
        
        Args:
            from_concept_id: Source concept UUID
            to_concept_id: Target concept UUID
            
        Returns:
            True if association removed successfully
        """
        try:
            await self._request("DELETE", f"/associations/{from_concept_id}/{to_concept_id}")
            return True
        except LeafMindError as e:
            if "not found" in str(e).lower():
                return False
            raise
    
    # Recall Operations
    async def recall_from_concept(
        self, 
        concept_id: str, 
        query: Optional[RecallQuery] = None
    ) -> List[RecallResult]:
        """
        Recall memories starting from a specific concept.
        
        Args:
            concept_id: The UUID of the source concept
            query: Optional recall query configuration
            
        Returns:
            List of recall results
        """
        if query is None:
            query = RecallQuery()
            
        request_data = {
            "source_concept_id": concept_id,
            "max_results": query.max_results,
            "min_relevance": query.min_relevance,
            "use_recency_boost": query.use_recency_boost,
            "exploration_breadth": query.exploration_breadth,
            "max_path_length": query.max_path_length
        }
        
        response = await self._request("POST", "/recall", request_data)
        return [
            RecallResult(
                concept=Concept(**result["concept"]),
                relevance_score=result["relevance_score"],
                path_length=result["path_length"],
                connection_strength=result["connection_strength"]
            )
            for result in response["results"]
        ]
    
    async def recall_by_content(
        self, 
        content_query: str, 
        query: Optional[RecallQuery] = None
    ) -> List[RecallResult]:
        """
        Recall memories by content similarity.
        
        Args:
            content_query: Text to search for
            query: Optional recall query configuration
            
        Returns:
            List of recall results
        """
        if query is None:
            query = RecallQuery()
            
        request_data = {
            "content_query": content_query,
            "max_results": query.max_results,
            "min_relevance": query.min_relevance,
            "use_recency_boost": query.use_recency_boost,
            "exploration_breadth": query.exploration_breadth,
            "max_path_length": query.max_path_length
        }
        
        response = await self._request("POST", "/recall/content", request_data)
        return [
            RecallResult(
                concept=Concept(**result["concept"]),
                relevance_score=result["relevance_score"],
                path_length=result["path_length"],
                connection_strength=result["connection_strength"]
            )
            for result in response["results"]
        ]
    
    async def spreading_activation_recall(self, concept_id: str) -> List[RecallResult]:
        """
        Perform spreading activation recall from a concept.
        
        Args:
            concept_id: The UUID of the source concept
            
        Returns:
            List of recall results
        """
        request_data = {"source_concept_id": concept_id}
        response = await self._request("POST", "/recall/spreading", request_data)
        return [
            RecallResult(
                concept=Concept(**result["concept"]),
                relevance_score=result["relevance_score"],
                path_length=result["path_length"],
                connection_strength=result["connection_strength"]
            )
            for result in response["results"]
        ]
    
    # Memory Management
    async def consolidate_memory(self) -> Dict[str, Any]:
        """Consolidate memory (move short-term to long-term)."""
        return await self._request("POST", "/memory/consolidate")
    
    async def forget_memories(self) -> Dict[str, Any]:
        """Forget weak memories (cleanup)."""
        return await self._request("POST", "/memory/forget")
    
    async def optimize_memory(self) -> Dict[str, Any]:
        """Optimize memory (consolidate + forget)."""
        return await self._request("POST", "/memory/optimize")
    
    # Persistence Operations
    async def save_to_disk(self) -> Dict[str, Any]:
        """Save current memory state to disk."""
        request_data = {"action": "Save"}
        return await self._request("POST", "/persistence", request_data)
    
    async def load_from_disk(self) -> Dict[str, Any]:
        """Load memory state from disk."""
        request_data = {"action": "Load"}
        return await self._request("POST", "/persistence", request_data)
    
    async def backup_database(self, backup_path: str) -> Dict[str, Any]:
        """Create a backup of the database."""
        request_data = {"action": {"Backup": {"path": backup_path}}}
        return await self._request("POST", "/persistence", request_data)
    
    async def restore_database(self, backup_path: str) -> Dict[str, Any]:
        """Restore database from a backup."""
        request_data = {"action": {"Restore": {"path": backup_path}}}
        return await self._request("POST", "/persistence", request_data)
    
    async def optimize_storage(self) -> Dict[str, Any]:
        """Optimize database storage."""
        request_data = {"action": "Optimize"}
        return await self._request("POST", "/persistence", request_data)
    
    async def get_persistence_stats(self) -> Optional[Dict[str, Any]]:
        """Get persistence statistics."""
        return await self._request("GET", "/persistence/stats")
    
    # Batch Operations
    async def batch_learn(self, concepts: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
        """
        Learn multiple concepts in a batch.
        
        Args:
            concepts: List of concept dictionaries with 'content' and optional 'metadata'
            
        Returns:
            List of responses for each concept
        """
        return await self._request("POST", "/batch/learn", concepts)
    
    async def batch_associate(self, associations: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
        """
        Create multiple associations in a batch.
        
        Args:
            associations: List of association dictionaries
            
        Returns:
            List of responses for each association
        """
        return await self._request("POST", "/batch/associate", associations)


class LeafMindError(Exception):
    """Exception raised for LeafMind API errors."""
    pass


# Convenience functions for common patterns
async def quick_learn_and_associate(
    client: LeafMindClient, 
    concepts: List[str], 
    create_associations: bool = True
) -> List[str]:
    """
    Quickly learn multiple concepts and optionally create associations between them.
    
    Args:
        client: LeafMind client instance
        concepts: List of concept contents to learn
        create_associations: Whether to create bidirectional associations between all concepts
        
    Returns:
        List of concept IDs
    """
    # Learn all concepts
    concept_ids = []
    for content in concepts:
        concept_id = await client.learn(content)
        concept_ids.append(concept_id)
    
    # Create associations if requested
    if create_associations and len(concept_ids) > 1:
        for i, from_id in enumerate(concept_ids):
            for j, to_id in enumerate(concept_ids):
                if i != j:
                    await client.associate(from_id, to_id, bidirectional=False)
    
    return concept_ids


async def build_knowledge_graph(
    client: LeafMindClient, 
    knowledge_pairs: List[tuple]
) -> Dict[str, str]:
    """
    Build a knowledge graph from concept pairs.
    
    Args:
        client: LeafMind client instance
        knowledge_pairs: List of (concept1, concept2) tuples to associate
        
    Returns:
        Dictionary mapping concept content to concept IDs
    """
    concept_map = {}
    
    # Learn all unique concepts
    all_concepts = set()
    for concept1, concept2 in knowledge_pairs:
        all_concepts.add(concept1)
        all_concepts.add(concept2)
    
    for concept in all_concepts:
        concept_id = await client.learn(concept)
        concept_map[concept] = concept_id
    
    # Create associations
    for concept1, concept2 in knowledge_pairs:
        await client.associate(
            concept_map[concept1], 
            concept_map[concept2], 
            bidirectional=True
        )
    
    return concept_map